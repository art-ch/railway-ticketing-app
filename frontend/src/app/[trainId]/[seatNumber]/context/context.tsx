'use client';

import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState
} from 'react';
import { useTrainPageContext } from '../../context';
import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { axiosInstance } from '@/api/axios/axios';

/** Using `undefined` as initial state to indicate "not yet fetched" */
export type BookingData = BookingType | null | undefined;

export type BookingPageContextProps = {
  bookingData: BookingData;
  isBookingLoading: boolean;
  updateBooking: (
    seatNumber: number,
    bookingId?: string,
    newBooking?: BookingType | null
  ) => void;
};

export const BookingPageContext = createContext<BookingPageContextProps | null>(
  null
);

export type BookingPageProviderProps = {
  children: React.ReactNode;
  seatNumber: string;
};

export const BookingPageProvider = ({
  children,
  seatNumber
}: BookingPageProviderProps) => {
  const { trainData, setTrainData } = useTrainPageContext();

  const [bookingData, setBookingData] = useState<BookingData>(undefined);
  const [isBookingLoading, setIsBookingLoading] = useState(false);

  const fetchBooking = useCallback(
    async (bookingId: string) => {
      setIsBookingLoading(true);

      try {
        const response = await axiosInstance.get(`/bookings/${bookingId}`);
        setBookingData(response.data.data);
      } catch (error) {
        console.error(`Error fetching booking for seat ${seatNumber}:`, error);
        setBookingData(null);
      } finally {
        setIsBookingLoading(false);
      }
    },
    [seatNumber]
  );

  useEffect(() => {
    if (!trainData) return;

    const seatData = trainData.seats.find(
      (seat) => seat.seatNumber.toString() === seatNumber
    );

    if (!seatData) {
      setBookingData(null);
      return;
    }

    if (seatData.bookingId) {
      fetchBooking(seatData.bookingId);
    } else {
      // No booking for this seat
      setBookingData(null);
    }
  }, [fetchBooking, trainData, seatNumber]);

  const updateBooking = (
    seatNumber: number,
    bookingId?: string,
    newBooking?: BookingType | null
  ) => {
    // Update the train data
    updateTrainSeatData(seatNumber, bookingId);

    // Update the booking data based on the scenario
    handleBookingDataUpdate(bookingId, newBooking);
  };

  const updateTrainSeatData = (seatNumber: number, bookingId?: string) => {
    setTrainData((prevData) => {
      if (!prevData) return null;

      const updatedSeats = prevData.seats.map((seat) =>
        seat.seatNumber === seatNumber ? { ...seat, bookingId } : seat
      );

      return {
        ...prevData,
        seats: updatedSeats
      };
    });
  };

  const handleBookingDataUpdate = (
    bookingId?: string,
    newBooking?: BookingType | null
  ) => {
    // Case 1: Explicit booking data provided
    if (newBooking !== undefined) {
      setBookingData(newBooking);
      return;
    }

    // Case 2: New booking ID provided but no data
    if (bookingId) {
      fetchBooking(bookingId);
      return;
    }

    // Case 3: No booking ID (seat is now unbooked)
    setBookingData(null);
  };

  return (
    <BookingPageContext.Provider
      value={{
        bookingData,
        isBookingLoading,
        updateBooking
      }}
    >
      {children}
    </BookingPageContext.Provider>
  );
};

export const useBookingPageContext = () => {
  const context = useContext(BookingPageContext);

  if (!context) {
    throw new Error(
      'useBookingPageContext must be used within a TrainPageProvider'
    );
  }

  return context;
};
