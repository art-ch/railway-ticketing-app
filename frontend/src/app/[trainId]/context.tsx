'use client';
import { createContext, useContext, useState } from 'react';

import { SetState } from '@/types/types';
import { Train } from 'railway-ticketing-app-sdk';

export type TrainPageData = Train | null;

export type TrainPageContextProps = {
  trainData: TrainPageData;
  setTrainData: SetState<TrainPageData>;
};

export const TrainPageContext = createContext<TrainPageContextProps | null>(
  null
);

export type TrainPageProviderProps = {
  children: React.ReactNode;
  pageData: TrainPageData;
};

export const TrainPageProvider = ({
  children,
  pageData
}: TrainPageProviderProps) => {
  const [trainData, setTrainData] = useState<TrainPageData>(pageData);

  return (
    <TrainPageContext.Provider
      value={{
        trainData,
        setTrainData
      }}
    >
      {children}
    </TrainPageContext.Provider>
  );
};
export const useTrainPageContext = () => {
  const context = useContext(TrainPageContext);

  if (!context) {
    throw new Error(
      'useTrainPageContext must be used within a TrainPageProvider'
    );
  }

  return context;
};
