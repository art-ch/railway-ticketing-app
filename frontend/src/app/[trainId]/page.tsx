import { Metadata } from 'next';
import { Train } from '@/components/pages/Train/Train';
import { fetchTrain } from './utils/utils';

export type TrainPageProps = { params: { trainId: string } };

export async function generateMetadata({
  params
}: TrainPageProps): Promise<Metadata> {
  // await is required here. Details at https://nextjs.org/docs/messages/sync-dynamic-apis
  const { trainId } = await params;

  return {
    title: `Train | ${trainId}`,
    description: `Information about train ${trainId}`
  };
}
export default async function TrainPage({ params }: TrainPageProps) {
  // await is required here. Details at https://nextjs.org/docs/messages/sync-dynamic-apis
  const { trainId } = await params;

  const trainInfo = await fetchTrain(trainId);

  return <Train trainInfo={trainInfo} />;
}
