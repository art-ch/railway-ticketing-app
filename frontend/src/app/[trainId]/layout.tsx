import { LayoutProps } from '@/types/types';
import { getFormattedTime } from '@/utils/formatters';
import { Metadata } from 'next';
import { TrainPageProvider } from './context';
import { fetchTrain } from './utils/utils';
import { notFound } from 'next/navigation';

export type TrainPageMetadataProps = { params: { trainId: string } };

export async function generateMetadata({
  params
}: TrainPageMetadataProps): Promise<Metadata> {
  // await is required here. Details at https://nextjs.org/docs/messages/sync-dynamic-apis
  const { trainId } = await params;

  return {
    title: `Train | ${trainId}`,
    description: `Information about train ${trainId}`
  };
}

export type TrainPageLayoutProps = LayoutProps & {
  params: { trainId: string };
};

export default async function TrainPageLayout({
  children,
  params
}: TrainPageLayoutProps) {
  const { trainId } = await params;

  const trainInfo = await fetchTrain(trainId);

  if (!trainInfo) {
    notFound();
  }

  const { departureStation, arrivalStation, departureTime, name, trainType } =
    trainInfo;

  return (
    <div>
      <div className="w-full border-b border-slate-300 p-2">
        <h1>
          {departureStation} â†’ {arrivalStation},{' '}
          {getFormattedTime(new Date(departureTime))}
        </h1>

        <div className="flex gap-2">
          <div>{name}</div>
          <div>{trainType}</div>
        </div>
      </div>

      <TrainPageProvider pageData={trainInfo}>{children}</TrainPageProvider>
    </div>
  );
}
