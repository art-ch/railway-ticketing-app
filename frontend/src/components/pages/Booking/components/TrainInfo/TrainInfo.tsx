'use client';

import React from 'react';

import { useTrainPageContext } from '@/app/[trainId]/context';
import { getFormattedTime } from '@/utils/formatters';

export const TrainInfo = () => {
  const { trainData: trainInfo } = useTrainPageContext();

  if (!trainInfo) {
    return null;
  }

  const {
    trainType,
    departureStation,
    arrivalStation,
    departureTime,
    arrivalTime
  } = trainInfo;

  return (
    <div>
      <p className="font-bold pb-1">Train Info</p>
      <table className="w-full">
        <thead>
          <tr className="text-sm font-semibold">
            <th className="text-left w-[100px]">Type</th>
            <th className="text-left w-[175px]">Origin</th>
            <th className="text-left w-[175px]">Destination</th>
            <th className="text-left w-[50px]">Dep.</th>
            <th className="text-left w-[50px]">Arr.</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>{trainType}</td>
            <td>{departureStation}</td>
            <td>{arrivalStation}</td>
            <td>{getFormattedTime(new Date(departureTime))}</td>
            <td>{getFormattedTime(new Date(arrivalTime))}</td>
          </tr>
        </tbody>
      </table>
    </div>
  );
};
