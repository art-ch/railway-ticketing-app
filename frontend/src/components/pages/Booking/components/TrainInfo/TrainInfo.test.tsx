import { render } from '@testing-library/react';
import { TrainInfo } from './TrainInfo';
import { useTrainPageContext } from '@/app/[trainId]/context';

jest.mock('@/app/[trainId]/context', () => ({
  useTrainPageContext: jest.fn().mockReturnValue({
    trainData: {
      trainId: '1',
      seats: [{ seatNumber: 1, isBooked: false }],
      trainType: 'local',
      name: 'train-name',
      departureStation: 'Kyiv',
      arrivalStation: 'Odesa',
      departureTime: '2025-03-31T05:05:11Z',
      arrivalTime: '2025-03-31T15:05:11Z'
    }
  })
}));

describe('BookingDataSkeleton component', () => {
  it('should be rendered correctly', () => {
    const { container } = render(<TrainInfo />);

    expect(container).toMatchSnapshot();
  });

  it('should be rendered correctly when there is no data', () => {
    (useTrainPageContext as jest.Mock).mockReturnValue({ trainData: null });

    const { container } = render(<TrainInfo />);

    expect(container).toMatchSnapshot();
  });
});
