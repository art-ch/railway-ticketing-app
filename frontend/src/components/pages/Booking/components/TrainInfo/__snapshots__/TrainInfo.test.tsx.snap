// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDataSkeleton component should be rendered correctly 1`] = `
<div>
  <div>
    <p
      class="font-bold pb-1"
    >
      Train Info
    </p>
    <table
      class="w-full"
    >
      <thead>
        <tr
          class="text-sm font-semibold"
        >
          <th
            class="text-left w-[100px]"
          >
            Type
          </th>
          <th
            class="text-left w-[175px]"
          >
            Origin
          </th>
          <th
            class="text-left w-[175px]"
          >
            Destination
          </th>
          <th
            class="text-left w-[50px]"
          >
            Dep.
          </th>
          <th
            class="text-left w-[50px]"
          >
            Arr.
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            local
          </td>
          <td>
            Kyiv
          </td>
          <td>
            Odesa
          </td>
          <td>
            08:05
          </td>
          <td>
            18:05
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`BookingDataSkeleton component should be rendered correctly when there is no data 1`] = `<div />`;
